gax/src/main/java/com/google/api/gax/batching/BatcherStats.java

```diff
@@ -0,0 +1,181 @@
+   * count of the entries themselves.
+   *
+   * <p>Note: This method aggregates all the subclasses of {@link ApiException} under ApiException
+   * using the {@link Code status codes} and its number of occurrences.
+   */
+  synchronized <T extends ApiFuture> void recordBatchElementsCompletion(
+      List<T> batchElementResultFutures) {
+    boolean isRequestPartiallyFailed = false;
+    for (final ApiFuture elementResult : batchElementResultFutures) {
+      try {
+        elementResult.get();
+      } catch (Throwable throwable) {
+
+        if (!isRequestPartiallyFailed) {
+          isRequestPartiallyFailed = true;
+          requestPartialFailureCount++;
+        }
+        Throwable actualCause = throwable.getCause();
+        Class exceptionClass = actualCause.getClass();
+
+        if (actualCause instanceof ApiException) {
+          Code code = ((ApiException) actualCause).getStatusCode().getCode();
+          exceptionClass = ApiException.class;
+
+          int oldStatusCount = MoreObjects.firstNonNull(entryStatusCounts.get(code), 0);
+          entryStatusCounts.put(code, oldStatusCount + 1);
+        }
+
+        int oldExceptionCount =
+            MoreObjects.firstNonNull(entryExceptionCounts.get(exceptionClass), 0);
+        entryExceptionCounts.put(exceptionClass, oldExceptionCount + 1);
+      }
+    }
+  }
+
+  /** Calculates and formats the message with request and entry failure count. */
+  @Nullable
+  synchronized BatchingException asException() {
+    if (requestExceptionCounts.isEmpty() && requestPartialFailureCount == 0) {
+      return null;
+    }
+
+    StringBuilder sb = new StringBuilder();
+    sb.append("Batching finished with ");
+
+    if (!requestExceptionCounts.isEmpty()) {
+      sb.append(String.format("%d batches failed to apply due to: ", requestExceptionCounts.size()))
+          .append(printKeyValue(requestExceptionCounts, requestStatusCounts))
+          .append(" and ");
+    }
+
+    sb.append(String.format("%d partial failures.", requestPartialFailureCount));
+    if (requestPartialFailureCount > 0) {
+      int totalEntriesCount = 0;
+      for (Integer count : entryExceptionCounts.values()) {
+        totalEntriesCount += count;
+      }
+
+      sb.append(
+              String.format(
+                  " The %d partial failures contained %d entries that failed with: ",
+                  requestPartialFailureCount, totalEntriesCount))
+          .append(printKeyValue(entryExceptionCounts, entryStatusCounts))
+          .append(".");
+    }
+    return new BatchingException(sb.toString());
+  }
+
+  /**
+   * Prints the class name and it's count along with {@link Code status code} and it's counts.
+   *
+   * <p>Example: "1 IllegalStateException, 1 ApiException(1 UNAVAILABLE, 1 ALREADY_EXISTS)".
+   */
+  private String printKeyValue(
+      Map<Class, Integer> exceptionCounts, Map<Code, Integer> statusCounts) {
+    StringBuilder keyValue = new StringBuilder();
```

Comment:
Naming is hard, but `keyValue` doesn't really describe this variable. Maybe `messageBuilder`?

Reply:
Thanks for pointing this out, Have updated the same.